cpp/include/experimental/graph_view.hpp

```diff
@@ -0,0 +1,274 @@
+               bool is_weighted)
+    : handle_ptr_(&handle),
+      number_of_vertices_(number_of_vertices),
+      number_of_edges_(number_of_edges),
+      properties_({is_symmetric, is_multigraph, is_weighted}){};
+
+  vertex_t get_number_of_vertices() const { return number_of_vertices_; }
+  edge_t get_number_of_edges() const { return number_of_edges_; }
+
+  bool is_symmetric() const { return properties_.is_symmetric; }
+  bool is_multigraph() const { return properties_.is_multigraph; }
+  bool is_weighted() const { return properties_.is_weighted; }
+
+ protected:
+  raft::handle_t const* get_handle_ptr() const { return handle_ptr_; };
+
+ private:
+  raft::handle_t const* handle_ptr_{nullptr};
+
+  vertex_t number_of_vertices_{0};
+  edge_t number_of_edges_{0};
+
+  graph_properties_t properties_{};
+};
+
+}  // namespace detail
+
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          bool store_transposed,
+          bool multi_gpu,
+          typename Enable = void>
+class graph_view_t;
+
+// multi-GPU version
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          bool store_transposed,
+          bool multi_gpu>
+class graph_view_t<vertex_t,
+                   edge_t,
+                   weight_t,
+                   store_transposed,
+                   multi_gpu,
+                   std::enable_if_t<multi_gpu>>
+  : public detail::graph_base_t<vertex_t, edge_t, weight_t> {
+ public:
+  using vertex_type                              = vertex_t;
+  using edge_type                                = edge_t;
+  using weight_type                              = weight_t;
+  static constexpr bool is_adj_matrix_transposed = store_transposed;
+  static constexpr bool is_multi_gpu             = multi_gpu;
+
+  graph_view_t(raft::handle_t const& handle,
+               std::vector<edge_t const*> const& adj_matrix_partition_offsets,
+               std::vector<vertex_t const*> const& adj_matrix_partition_indices,
+               std::vector<weight_t const*> const& adj_matrix_partition_weights,
+               std::vector<vertex_t> const& vertex_partition_segment_offsets,
+               partition_t<vertex_t> const& partition,
+               vertex_t number_of_vertices,
+               edge_t number_of_edges,
+               bool is_symmetric,
+               bool is_multigraph,
+               bool is_weighted,
+               bool sorted_by_global_degree_within_vertex_partition,
+               bool do_expensive_check = false);
+
+  vertex_t get_number_of_local_vertices() const
+  {
+    auto comm_p_rank = this->get_handle_ptr()->get_comms().get_rank();
+    return partition_.vertex_partition_offsets[comm_p_rank + 1] -
+           partition_.vertex_partition_offsets[comm_p_rank];
+  }
+
+  size_t get_number_of_adj_matrix_partitions() { return adj_matrix_partition_offsets_.size(); }
+
+  // Better avoid direct invocation in application code.
```

Comment:
Yeah... this makes sense, but the problem is the legacy code. They currently directly accesses offsets, indices, and weights arrays.

I will update comments to better explain this and how this will change in the future.

Updated the comments to reflect this.

// FIXME: this function is not part of the public stable API.This function is mainly for pattern
  // accelerator implementation. This function is currently public to support the legacy
  // implementations directly accessing CSR/CSC data, but this function will eventually become
  // private or even disappear if we switch to CSR + DCSR (or CSC + DCSC).